File: examples/js-sdk/firewing-sdk.ts
/**
 * FIREWING JS/TS SDK.
 *
 * Minimal fetch-based client — no dependencies. Works in Node 18+
 * (built-in fetch) and modern browsers.
 *
 * Usage:
 *   import { FirewingClient } from "./firewing-sdk";
 *   const client = new FirewingClient("http://localhost:8000", apiKey);
 *   const reply = await client.chat([{ role: "user", content: "Hello" }]);
 *   for await (const chunk of client.chatStream([{ role: "user", content: "Hello" }])) {
 *     process.stdout.write(chunk);
 *   }
 */

export interface ChatMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface ChatOptions {
  model?: string;
  temperature?: number;
  topP?: number;
  maxTokens?: number;
  persona?: string;
}

export class FirewingError extends Error {
  requestId?: string;
  constructor(message: string, requestId?: string) {
    super(message);
    this.name = "FirewingError";
    this.requestId = requestId;
  }
}

export class FirewingClient {
  private baseUrl: string;
  private apiKey: string;

  constructor(baseUrl: string, apiKey: string) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.apiKey = apiKey;
  }

  private headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
    };
  }

  async health(): Promise<{ status: string; version: string }> {
    const resp = await fetch(`${this.baseUrl}/health`);
    if (!resp.ok) throw new FirewingError(`HTTP ${resp.status}`);
    return resp.json();
  }

  async chat(messages: ChatMessage[], opts: ChatOptions = {}): Promise<string> {
    const resp = await fetch(`${this.baseUrl}/v1/chat/completions`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        model: opts.model ?? "firewing-1.0-beta",
        messages,
        temperature: opts.temperature ?? 0.7,
        top_p: opts.topP ?? 0.9,
        max_tokens: opts.maxTokens ?? 1024,
        stream: false,
        persona: opts.persona ?? null,
      }),
    });
    if (!resp.ok) await this.throwForError(resp);
    const data = await resp.json();
    return data.choices[0].message.content;
  }

  async *chatStream(messages: ChatMessage[], opts: ChatOptions = {}): AsyncGenerator<string> {
    const resp = await fetch(`${this.baseUrl}/v1/chat/completions`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({
        model: opts.model ?? "firewing-1.0-beta",
        messages,
        temperature: opts.temperature ?? 0.7,
        top_p: opts.topP ?? 0.9,
        max_tokens: opts.maxTokens ?? 1024,
        stream: true,
        persona: opts.persona ?? null,
      }),
    });
    if (!resp.ok || !resp.body) await this.throwForError(resp);

    const reader = resp.body!.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n\n");
      buffer = lines.pop() ?? "";
      for (const line of lines) {
        if (!line.startsWith("data: ")) continue;
        const payload = line.slice("data: ".length);
        if (payload === "[DONE]") return;
        yield payload;
      }
    }
  }

  private async throwForError(resp: Response): Promise<never> {
    let message = `HTTP ${resp.status}`;
    let requestId: string | undefined;
    try {
      const data = await resp.json();
      if (data?.error) {
        message = data.error.message ?? message;
        requestId = data.error.request_id;
      }
    } catch {
      // response wasn't JSON — fall back to the status-only message
    }
    throw new FirewingError(message, requestId);
  }
}

File: examples/python-sdk/firewing_sdk.py
"""FIREWING Python SDK.

Minimal client for the FIREWING API — no external dependencies beyond
`requests`, deliberately kept small rather than wrapping the OpenAI SDK
(which would imply a compatibility guarantee this beta doesn't make
yet).

Usage:
    from firewing_sdk import FirewingClient

    client = FirewingClient(base_url="http://localhost:8000", api_key="...")
    reply = client.chat([{"role": "user", "content": "Hello"}])
    print(reply)

    for chunk in client.chat_stream([{"role": "user", "content": "Hello"}]):
        print(chunk, end="", flush=True)
"""

from __future__ import annotations

import json
from typing import Iterator

import requests


class FirewingError(RuntimeError):
    def __init__(self, message: str, request_id: str | None = None):
        super().__init__(message)
        self.request_id = request_id


class FirewingClient:
    def __init__(self, base_url: str, api_key: str, timeout: int = 120):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    def _headers(self) -> dict:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    def health(self) -> dict:
        resp = requests.get(f"{self.base_url}/health", timeout=self.timeout)
        resp.raise_for_status()
        return resp.json()

    def chat(
        self,
        messages: list[dict],
        model: str = "firewing-1.0-beta",
        temperature: float = 0.7,
        top_p: float = 0.9,
        max_tokens: int = 1024,
        persona: str | None = None,
    ) -> str:
        body = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "top_p": top_p,
            "max_tokens": max_tokens,
            "stream": False,
            "persona": persona,
        }
        resp = requests.post(
            f"{self.base_url}/v1/chat/completions",
            headers=self._headers(),
            json=body,
            timeout=self.timeout,
        )
        if not resp.ok:
            self._raise_for_error(resp)
        data = resp.json()
        return data["choices"][0]["message"]["content"]

    def chat_stream(
        self,
        messages: list[dict],
        model: str = "firewing-1.0-beta",
        temperature: float = 0.7,
        top_p: float = 0.9,
        max_tokens: int = 1024,
        persona: str | None = None,
    ) -> Iterator[str]:
        body = {
            "model": model,
            "messages": messages,
            "temperature": temperature,
            "top_p": top_p,
            "max_tokens": max_tokens,
            "stream": True,
            "persona": persona,
        }
        with requests.post(
            f"{self.base_url}/v1/chat/completions",
            headers=self._headers(),
            json=body,
            timeout=self.timeout,
            stream=True,
        ) as resp:
            if not resp.ok:
                self._raise_for_error(resp)
            for line in resp.iter_lines(decode_unicode=True):
                if not line or not line.startswith("data: "):
                    continue
                payload = line[len("data: "):]
                if payload == "[DONE]":
                    break
                yield payload

    def _raise_for_error(self, resp: requests.Response) -> None:
        try:
            err = resp.json().get("error", {})
            raise FirewingError(
                err.get("message", resp.text), request_id=err.get("request_id")
            )
        except (json.JSONDecodeError, ValueError):
            raise FirewingError(f"HTTP {resp.status_code}: {resp.text}")
